blueflood-core/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/RollupHandler.java

```diff
@@ -56,114 +63,145 @@
-
-        // if Granularity is FULL, we are missing raw data - can't generate that
-        if (ROLLUP_REPAIR && isRollable && g != Granularity.FULL && metricData != null) {
-            final Timer.Context rollupsCalcCtx = rollupsCalcOnReadTimer.time();
-
-            if (metricData.getData().isEmpty()) { // data completely missing for range. complete repair.
-                rollupsRepairEntireRange.mark();
-                List<Points.Point> repairedPoints = repairRollupsOnRead(locator, g, from, to);
-                for (Points.Point repairedPoint : repairedPoints) {
-                    metricData.getData().add(repairedPoint);
-                }
-
-                if (repairedPoints.isEmpty()) {
-                    rollupsRepairEntireRangeEmpty.mark();
-                }
-            } else {
-                long actualStart = minTime(metricData.getData());
-                long actualEnd = maxTime(metricData.getData());
-
-                // If the returned start is greater than 'from', we are missing a portion of data.
-                if (actualStart > from) {
-                    rollupsRepairedLeft.mark();
-                    List<Points.Point> repairedLeft = repairRollupsOnRead(locator, g, from, actualStart);
-                    for (Points.Point repairedPoint : repairedLeft) {
-                        metricData.getData().add(repairedPoint);
-                    }
-
-                    if (repairedLeft.isEmpty()) {
-                        rollupsRepairedLeftEmpty.mark();
-                    }
-                }
-
-                // If the returned end timestamp is less than 'to', we are missing a portion of data.
-                if (actualEnd + g.milliseconds() <= to) {
-                    rollupsRepairedRight.mark();
-                    List<Points.Point> repairedRight = repairRollupsOnRead(locator, g, actualEnd + g.milliseconds(), to);
-                    for (Points.Point repairedPoint : repairedRight) {
-                        metricData.getData().add(repairedPoint);
-                    }
-
-                    if (repairedRight.isEmpty()) {
-                        rollupsRepairedRightEmpty.mark();
-                    }
-                }
-            }
-            rollupsCalcCtx.stop();
+        // Now do rollups on read asynchronously
+        ArrayList<ListenableFuture<Boolean>> futures = new ArrayList<ListenableFuture<Boolean>>();
+        for (final Map.Entry<Locator, MetricData> metricData: metricDataMap.entrySet()) {
+            futures.add(
+                RollupsOnReadExecutor.submit(new Callable<Boolean>() {
+                    @Override
+                    public Boolean call() {
+                            boolean isRollable = metricData.getValue().getType().equals(MetricData.Type.NUMBER.toString())
+                                    || metricData.getValue().getType().equals(MetricData.Type.HISTOGRAM.toString());
+                            Boolean retValue = false;
```

Comment:
Yes, we need the `Boolean` variable because `Callable` always returns an object.